### PR TITLE
Refs #24840 -- Added GDALRaster warp and transform methods.

### PR DESCRIPTION
**Pull request by yellowcap**
- The warp method allows altering any parameters of the raster (geotranform, datatype, srid, etc
- The transform method allows reprojecting the raster into a different srid

### CODE REVIEW

File: django/contrib/gis/gdal/raster/source.py [hunk 1]
```diff
@@ -278,3 +280,122 @@ def bands(self):
...
+        By default, the warp functions keeps all parameters equal to the values
+        of the original source raster. For the name of the target raster, the
+        name of the source raster will be used and appended with
+        _copy. + source_driver_name.
+
+        In addition, the resampling algorithm can be specified with the "resampling"
+        input parameter. The default is NearestNeighbor. For a list of all options
+        consult the GDAL_RESAMPLE_ALGORITHMS constant.
+        """
+        # Get the parameters defining the geotransform, srid and size of the raster
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/source.py]: , and

File: django/contrib/gis/gdal/raster/source.py [hunk 2]
```diff
@@ -278,3 +280,122 @@ def bands(self):
...
+        if 'origin' not in ds_input:
+            ds_input['origin'] = self.origin
+
+        if 'scale' not in ds_input:
+            ds_input['scale'] = self.scale
+
+        if 'skew' not in ds_input:
+            ds_input['skew'] = self.skew
+
+        # Get the driver, name and datatype of the target raster
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/source.py]: , and

File: django/contrib/gis/gdal/raster/source.py [hunk 3]
```diff
@@ -278,3 +280,122 @@ def bands(self):
...
+
+        # Assure the signs of the scale are the same as in the source raster
+        scalex = copysign(scalex, self.scale.x)
+        scaley = copysign(scaley, self.scale.y)
+
+        # Construct the target warp dictionary
+        data = {
+            'origin': [origin.x, origin.y],
+            'scale': [scalex, scaley],
+            'srid': srid
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/source.py]: srid,

File: docs/ref/contrib/gis/gdal.txt [hunk 4]
```diff
@@ -1288,6 +1288,83 @@ blue.
             >>> rst.bands[1].data()
             array([[ 2.,  3.]], dtype=float32)
 
+    .. method:: warp(ds_input, resampling='NearestNeighbour')
+
+        .. versionadded:: 1.9
+
+        Returns a warped version of this raster.
+        
+        The warping parameters can be specified through the `ds_input` argument.
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: Need double backticks to have an effect on the rendered docs.

File: docs/ref/contrib/gis/gdal.txt [hunk 5]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        The use of `ds_input` is analogue to the corresponding argument of the
+        class constructor. It is a dictionary with the characteristics of the 
+        target raster. Allowed dictionary key values are width, height, SRID,
+        origin, scale, skew, datatype, driver, and name (filename).
+
+        By default, the warp functions keeps most parameters equal to the values
+        of the original source raster. So only parameters that should be changed
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: , so only...

File: docs/ref/contrib/gis/gdal.txt [hunk 6]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        need to be specified. Note that this includes the driver, so for file based
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: file-based

File: docs/ref/contrib/gis/gdal.txt [hunk 7]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        rasters the warp function will create a new raster on disk.
+        
+        The only parameter that is set differently from the source raster is the
+        name. The default value of the the raster name is the name of the source
+        raster appended with `_copy + source_driver_name`. For file based rasters
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: '_copy' + source_driver_name

File: docs/ref/contrib/gis/gdal.txt [hunk 8]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        it is recommended to provide the file path of the target raster.
+
+        The re-sampling algorithm used for warping can be specified with the
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: resampling

File: docs/ref/contrib/gis/gdal.txt [hunk 9]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        `resampling` argument. The default is `NearestNeighbor`, the other
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: , and the other

File: docs/ref/contrib/gis/gdal.txt [hunk 10]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+        to the `gdalwarp` command-line utility.
+
+        For example, the warp function can be used for aggregating a raster to the
+        double of its original pixel scale:
+
+            >>> rst = GDALRaster({
+            ...     "width": 6, "height": 6, "srid": 3086,
+            ...     "origin": [500000, 400000],
+            ...     "scale": [100, -100],
+            ...     "bands": [{"data": range(36), "nodata_value": 99}]
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: trailing comma

File: docs/ref/contrib/gis/gdal.txt [hunk 11]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+            ... })
+            >>> target = rst.warp({"scale": [200, -200], "width": 3, "height": 3})
+            >>> target.bands[0].data()
+            array([[  7.,   9.,  11.],
+                   [ 19.,  21.,  23.],
+                   [ 31.,  33.,  35.]], dtype=float32)
+
+    .. method:: transform(srid, driver=None, name=None, squared_pixels=True, resampling='NearestNeighbour'):
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: chop ending colon

File: docs/ref/contrib/gis/gdal.txt [hunk 12]
```diff
@@ -1288,6 +1288,83 @@ blue.
...
+
+        By default, the driver of the source raster is used and the name of the
+        raster is the original name appended with `_copy + source_driver_name`.
+        A different driver or name can be specified with the `driver` and `name`
+        arguments. The default re-sampling algorithm is `NearestNeighbour` but can
+        be changed using the `resampling` argument.
+
+        By default, the `x` and `y` scales of the pixels of the transformed raster
+        are set to be equal. The larger of the two values in the new projection is
+        used. To disable this behavior set the `squared_pixels` property to `False`.
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: behavior,

File: tests/gis_tests/gdal_tests/test_raster.py [hunk 13]
```diff
@@ -155,6 +155,101 @@ def test_file_based_raster_creation(self):
...
+            'width': 4,
+            'height': 4,
+            'nr_of_bands': 1,
+            'srid': 3086,
+            'origin': (500000, 400000),
+            'scale': (100, -100),
+            'skew': (0, 0),
+            'bands': [{
+                'data': range(16),
+                'nodata_value': 255
```
Comment (timgraham) [inline tests/gis_tests/gdal_tests/test_raster.py]: trailing comma here and next line

**Comment by timgraham** [comment]
Add 1.9 release notes?

**Comment by yellowcap** [comment]
Thanks for the review @timgraham . I added 1.9 release notes and included the corrections. However, Jenkins does not seem to like the word `resampling`, is there a way to add it to the spell checker? Otherwise we can just use `sampling` which would also work in this context.

**Comment by timgraham** [comment]
Add it to docs/spelling_wordlist.

**Comment by yellowcap** [comment]
I corrected the spelling issue and I rewrote the ``transform`` method to use the ``GDALAutoCreateWarpedVRT`` function in gdal. The result is very similar, but I think this method is more stable as it uses gdal to compute the transformed geotransform parameters. I think like this it will be more reliable overall.

**Comment by timgraham** [comment]
Some cosmetic and doc edits: http://dpaste.com/163BPM4
LGTM, pending @claudep's review of the functionality.

**Comment by claudep** [comment]
Great, committed in c0780215

**Pull request closed without merging**